imap/tls.c

```diff
@@ -1314,19 +1314,44 @@ EXPORTED int tls_start_servertls(int readfd, int writefd, int timeout,
     saslprops->ssf = (sasl_ssf_t) tls_cipher_usebits;
 
 #if (OPENSSL_VERSION_NUMBER >= 0x0090800fL)
-    if (SSL_session_reused(tls_conn)) {
-        saslprops->cbinding.len = SSL_get_finished(tls_conn,
-                                                   saslprops->tls_finished,
-                                                   MAX_FINISHED_LEN);
+    if (SSL_version(tls_conn) <= TLS1_2_VERSION) {
+        if (SSL_session_reused(tls_conn)) {
+            saslprops->cbinding.len = SSL_get_finished(tls_conn,
+                                                    saslprops->tls_finished,
+                                                    MAX_FINISHED_LEN);
+        }
+        else {
+            saslprops->cbinding.len = SSL_get_peer_finished(tls_conn,
+                                                    saslprops->tls_finished,
+                                                    MAX_FINISHED_LEN);
+        }
+
+        saslprops->cbinding.name = "tls-unique";
+        saslprops->cbinding.data = saslprops->tls_finished;
     }
+#if (OPENSSL_VERSION_NUMBER >= 0x10100001L)
     else {
-        saslprops->cbinding.len = SSL_get_peer_finished(tls_conn,
-                                                        saslprops->tls_finished,
-                                                        MAX_FINISHED_LEN);
+        /* see RFC 9266 Section 2, Definition of label, context, length */
+        static const char label[] = "EXPORTER-Channel-Binding";
+        static const size_t key_len = 32;
+
+        assert(key_len <= sizeof(saslprops->tls_finished));
```

Comment:
Since both key_len and tls_finished size (`MAX_FINISHED_LEN`) are constants, this looks like a good opportunity for `static_assert`. We can still fall back to `assert` if `#ifndef static_assert`

Reply:
Yes, `static_assert` is the best solution to compare constant values. However `static_assert` is only defined for C++11 and above. I think there is no benefit to replace a single line of code with additional compiler troubles like here https://stackoverflow.com/questions/3385515/static-assert-in-c

Reply:
`static_assert` is C++11/C23.  If we want to use it in Cyrus, it'll have to come with a bunch of back-compat baggage for portability -- which is fine if we're committed to making good use of it, but this isn't the PR for it.

My greater concern though is that Cyrus's `assert()` is very different from C's `assert()`, and this difference is already a source of confusion for casual contributors.  If we start using `static_assert` too (and especially if our fallback implementation of it relies on `assert()`), that will muddy the waters further.